### ISSUE TITLE
[FeatureRequest] Extend support for AssemblyDependencyResolver to NativeAoT and other platforms

### ISSUE THREAD

**Issue by ThaDaVos** [dotnet/sdk#42389]
### Is your feature request related to a problem? Please describe.
Currently the `System.Runtime.Loader.AssemblyDependencyResolver` seems to only support a few platforms and not NativeAoT.
Why? Cause I want to create a cross-platform application which is available on multiple platforms and I need plugin support as I want to use a module/plugin based architecture to prevent me needing to compile hundreds different versions of the application based on the modules used per version.

I followed the guide from https://learn.microsoft.com/en-us/dotnet/core/tutorials/creating-app-with-plugin-support how to set it up - but sadly, when compiling the `AppWithPlugin` project with NativeAoT I get the next error:
```
System.PlatformNotSupportedException: Operation is not supported on this platform.
   at System.Runtime.Loader.AssemblyDependencyResolver..ctor(String) + 0x24
   at AppWithPlugin.Program.LoadPlugin(String) + 0xe6
   at AppWithPlugin.Program.<>c.<Main>b__0_0(String) + 0xe
   at System.Linq.Enumerable.SelectManySingleSelectorIterator`2.ToList() + 0x70
   at AppWithPlugin.Program.Main(String[] args) + 0xfa
```
I guess this is due the following lines in `AssemblyDependencyResolver` also triggering for NativeAoT:
```C#
    [UnsupportedOSPlatform("android")]
    [UnsupportedOSPlatform("browser")]
    [UnsupportedOSPlatform("ios")]
    [UnsupportedOSPlatform("tvos")]
    public sealed class AssemblyDependencyResolver
```

### Describe the solution you'd like
Or for this to get supported - or be guided to an alternative way of creating a cross-platform plugin system/architecture with DotNet (Maui)

### Additional context
See also this question: https://github.com/dotnet/maui/discussions/5547

**Comment by vitek-karas** [comment]
@MichalStrehovsky 

NativeAOT currently doesn't have any way to dynamically load code, that's why it makes no sense to support `AssemblyDependencyResolver`, just like `Assembly.LoadFile` or `AssemblyLoadContext.Load.LoadFromAssemblyPath` are not supported either. There are several different pieces which make this very difficult:
* No way to run new IL - NativeAOT doesn't have a JIT or interpreter right now
* Trimming - the app is trimmed but that process assumes that it sees all the code that is ever going to part of the app. Adding more code dynamically breaks that assumption and would lead to runtime breaks.
* Assemblies are note files anymore - the normal assembly==file model which works with CoreCLR or .NET Framework doesn't work as well in NativeAOT or other "packaged" app formats (single-file is another example)

The MAUI side of things is similar and runs into similar set of issues. Some platforms do have a JIT or interpreter but not all configurations have that. The trimming problems are exactly the same. Also, the assembly==file assumption is somewhat weird on mobile due to the packaging requirements of the different platforms.

This doesn't mean we won't ever have this, but for now I am not aware of any plans to support this soon.

It would be interesting to understand what scenarios you're trying to solve with this, so that we can prioritize future work better.

**Comment by ThaDaVos** [comment]
The scenario I was/am needing this for is for a piece of software I want to develop which is used for ECR (Electronic Cash Registers) - I want to set this up modulary as not very branch/area needs every module - so I wanted to employ a plugin based system to create modules of functionaility which can be loaded into the program - but I want my ECR software to be available on the most common platforms, which is Windows, Android (mobile/tablets), MacOS (fancy people like dentists) and IoS (also mobile/tablet).

I was planning to go use Maui for this - and preferably NativeAOT to set this up - each plugin/module will have interfaces to implement and supply - I was thinking of somehow using Dependecy Injection for this too - so I can keep domains of the modules seperated.

Any idea what would be an alternative to using Libraries with code to create this? I know some old software used XML based content plugins to do this - but I rather not go there - also I rather know before hand if it will eventually become available before starting to develop as migrating later would be impossible

**Comment by ThaDaVos** [comment]
An alternative would be making everything Native compiled and use p/invoke and stuff - but I guess it would be impossible to add UI etc through that...

**Comment by MichalStrehovsky** [comment]
I think for this set up, you'd be better off with a JIT. Is there something in your requirements that makes JIT not suitable?

AOT doesn't generally mean better - it has some better characteristics, but also some drawbacks. It's possible a JIT would be able to do better in a dynamic setup like this (e.g. it can optimize code based on the modules and types that are loaded instead of pessimizing for the broadest possible option that AOT would need to do in an open system like this).

**Comment by vitek-karas** [comment]
> I think for this set up, you'd be better off with a JIT. 

While I do agree with this 100%, I don't think it helps with the original issue which is to do this in a MAUI app. The trimming will basically make any code-gen problems moot. MAUI can already do dynamic codegen on all platforms (perf problems aside), but the trimming will just not work - and I don't think untrimmed MAUI app is realistic, but that might be a thing to try if the size is not an issue.

**Comment by ThaDaVos** [comment]
Also - I thought Android and iOS by default do NativeAoT nowadays

**Comment by ThaDaVos** [comment]
So I can do - non-native and just use the plugin example - except that `AssemblyDependencyResolver` is not available on Android and iOS which I want to eventually target too - any solutions for that?

**Comment by vitek-karas** [comment]
`AssemblyDependencyResolver` is really just a helper to find all the assemblies on disk for the plugin. On mobile this would probably look differently, but in the end you can implement this yourself without much trouble.

**Comment by ThaDaVos** [comment]
Oh if that's the case - then I gonna see if I can find the code of the original one and see what I can create myself

**Comment by ThaDaVos** [comment]
Oh if that's the case - then I gonna see if I can find the code of the original one and see what I can create myself

**Comment by MasterMann** [comment]
Original source is here: https://github.com/dotnet/runtime/blob/d101ebb456012d979a20f497c13527d4bd91ad55/src/libraries/System.Private.CoreLib/src/System/Runtime/Loader/AssemblyDependencyResolver.cs#L17 @ThaDaVos 

I support this request as well.

**Comment by ThaDaVos** [comment]
> Original source is here: https://github.com/dotnet/runtime/blob/d101ebb456012d979a20f497c13527d4bd91ad55/src/libraries/System.Private.CoreLib/src/System/Runtime/Loader/AssemblyDependencyResolver.cs#L17 @ThaDaVos
> 
> I support this request as well.

Thank you! Hadn't had time to look for it - but I bet it will help a lot!

**Comment by MiYanni** [comment]
I know this is a feature request, but looking over the discussion, it is unlikely to be implemented into our `AssemblyDependencyResolver`. If you are still desiring development, please reopen/respond/create a new issue. Thanks!
